docarray/array/mixins/io/binary.py

```diff
@@ -41,37 +43,107 @@ def load_binary(
 
         with file_ctx as fp:
             d = fp.read() if hasattr(fp, 'read') else fp
-            if get_compress_ctx(algorithm=compress) is not None:
-                d = decompress_bytes(d, algorithm=compress)
-                compress = None
+        if get_compress_ctx(algorithm=compress) is not None:
+            d = decompress_bytes(d, algorithm=compress)
+            compress = None
 
-            if protocol == 'protobuf-array':
-                from ....proto.docarray_pb2 import DocumentArrayProto
+        if protocol == 'protobuf-array':
+            from ....proto.docarray_pb2 import DocumentArrayProto
 
-                dap = DocumentArrayProto()
-                dap.ParseFromString(d)
+            dap = DocumentArrayProto()
+            dap.ParseFromString(d)
 
-                return cls.from_protobuf(dap)
-            elif protocol == 'pickle-array':
-                return pickle.loads(d)
-            else:
-                _len = len(random_uuid().bytes)
-                _binary_delimiter = d[:_len]  # first get delimiter
-                if _show_progress:
-                    from rich.progress import track as _track
+            return cls.from_protobuf(dap)
+        elif protocol == 'pickle-array':
+            return pickle.loads(d)
 
-                    track = lambda x: _track(x, description='Deserializing')
-                else:
-                    track = lambda x: x
-                return cls(
-                    Document.from_bytes(od, protocol=protocol, compress=compress)
-                    for od in track(d[_len:].split(_binary_delimiter))
+        # Binary format for streaming case
+        else:
+
+            # 1 byte (uint8)
+            version = int.from_bytes(d[0:1], 'big', signed=False)
+            # 8 bytes (uint64)
+            num_docs = int.from_bytes(d[1:9], 'big', signed=False)
+
+            if show_progress:
+                from rich.progress import track as _track
+
+                track = lambda x: _track(x, description='Deserializing')
+            else:
+                track = lambda x: x
+
+            docs = []
+            start_pos = 9
+            for d in track(self):
+                # 4 bytes (uint32)
+                len_current_doc_in_bytes = int.from_bytes(
+                    d[start_pos : start_pos + 1], 'big', signed=False
                 )
+                start_doc_pos = start_pos + 1
+                end_doc_pos = start_doc_pos + len_current_doc_in_bytes
+                start_pos = end_doc_pos
+                # variable length bytes doc
+                doc = Document.from_bytes(
+                    d[start_doc_pos:end_doc_pos], protocol=protocol, compress=compress
+                )
+                docs.append(doc)
+
+            return cls(docs)
+
+    @classmethod
+    def _get_batches(cls, gen, batch_size):
+        for i in range(0, len(gen), batch_size):
+            yield gen[i : i + batch_size]
 
     @classmethod
     def from_bytes(
```

Comment:
Inspired from `pickle`,  the **protocol** in `.from_bytes()` can be automatically recognized from the binary file. So,  the user can use any protocol to dumps the DA via `.to_bytes(... protocol='...')`. Then, other users can dirrectly load the docs without knowing which protocol should be used.